src/zcash/prf.cpp

```diff
@@ -0,0 +1,63 @@
+#include "prf.h"
+#include "crypto/sha256.h"
+
+uint256 PRF(bool a, bool b, bool c, bool d,
+            const uint256& x,
+            const uint256& y)
+{
+    uint256 res;
+    unsigned char blob[64];
+
+    memcpy(&blob[0], x.begin(), 32);
+    memcpy(&blob[32], y.begin(), 32);
+
+    blob[0] &= 0x0F;
+    blob[0] |= (a ? 1 << 7 : 0) | (b ? 1 << 6 : 0) | (c ? 1 << 5 : 0) | (d ? 1 << 4 : 0);
```

Comment:
I want `PRF()` to throw an exception when the leading 4 bits aren't zero. That way you can verify it's collision resistant without looking at other code. It will also help prevent bugs where, for example, the commitment includes whatever random bits are at the start of a<sub>pk</sub> but they're ignored by everything else.